File: cpp/0109.0_Convert_Sorted_List_to_Binary_Search_Tree.cpp
/*
Runtime: 25 ms, faster than 80.67% of C++ online submissions for Convert Sorted List to Binary Search Tree.
Memory Usage: 28.5 MB, less than 47.40% of C++ online submissions for Convert Sorted List to Binary Search Tree.
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    vector<int> arr;
    
public:
    TreeNode* constructBST(int l, int r) {
        if (l > r) {
            return nullptr;
        }
        int mid = (l + r) / 2;
        return new TreeNode(arr[mid], constructBST(l, mid - 1), constructBST(mid + 1, r));
    }
    
    TreeNode* sortedListToBST(ListNode* head) {
        while (head != nullptr) {
            arr.push_back(head->val);
            head = head->next;
        }
        return constructBST(0, arr.size() - 1);        
    }
};

File: problems/0109.0_Convert_Sorted_List_to_Binary_Search_Tree.py
'''
DFS

Runtime: 136 ms, faster than 26.88% of Python3 online submissions for Convert Sorted List to Binary Search Tree.
Memory Usage: 20.4 MB, less than 10.50% of Python3 online submissions for Convert Sorted List to Binary Search Tree.
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
        vals = []
        while head:
            vals.append(head.val)
            head = head.next
        
        def construct(l: int, r: int) -> Optional[TreeNode]:
            if l > r:
                return None
            if l == r:
                return TreeNode(vals[l])
            mid = (l + r) // 2
            return TreeNode(vals[mid], construct(l, mid - 1), construct(mid + 1, r))
        
        return construct(0, len(vals) - 1)
        

'''
DFS
no need to use if l == r:

Runtime: 119 ms, faster than 88.17% of Python3 online submissions for Convert Sorted List to Binary Search Tree.
Memory Usage: 20.3 MB, less than 26.39% of Python3 online submissions for Convert Sorted List to Binary Search Tree.
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
        vals = []
        while head:
            vals.append(head.val)
            head = head.next
        
        def construct(l: int, r: int) -> Optional[TreeNode]:
            if l > r:
                return None
            mid = (l + r) // 2
            return TreeNode(vals[mid], construct(l, mid - 1), construct(mid + 1, r))
        
        return construct(0, len(vals) - 1)
        
